File: src/main/java/net/maple/packets/GuildPackets.java
package net.maple.packets;

import client.Character;
import net.maple.SendOpcode;
import util.packet.PacketWriter;

public class GuildPackets {

    public static void changeGuildName(Character chr, String name) {
        PacketWriter pw = new PacketWriter(8);

        pw.writeHeader(SendOpcode.USER_GUILD_NAME_CHANGED);
        pw.writeInt(chr.getId());
        pw.writeMapleString(name);

        chr.getField().broadcast(pw.createPacket(), chr);
    }

    public static void changeGuildMark(Character chr) {
        PacketWriter pw = new PacketWriter(8);

        pw.writeHeader(SendOpcode.USER_GUILD_MARK_CHANGED);
        pw.writeInt(chr.getId());
        pw.writeShort(0); // bg
        pw.write(0); // bg color
        pw.writeShort(0); // mark
        pw.write(0); // mark color

        chr.getField().broadcast(pw.createPacket(), chr);
    }

    public static void loadGuild(Character chr) {
        if (chr.getGuild() == null) return;

        PacketWriter pw = new PacketWriter(32);

        pw.writeHeader(SendOpcode.GUILD_RESULT);
        pw.writeByte(GuildRes.LOAD_GUILD_DONE);
        pw.writeBool(true);
        chr.getGuild().encode(pw);

        chr.write(pw.createPacket());
    }

    public static void expel(Character chr) {
        PacketWriter pw = new PacketWriter(4);

        pw.writeHeader(SendOpcode.GUILD_RESULT);
        pw.writeByte(GuildRes.LOAD_GUILD_DONE);
        pw.writeBool(false);

        chr.write(pw.createPacket());
    }

    public static final class GuildReq {
        public static final byte LOAD_GUILD = 0x0;
        public static final byte INPUT_GUILD_NAME = 0x1;
        public static final byte CHECK_GUILD_NAME = 0x2;
        public static final byte CREATE_GUILD_AGREE = 0x3;
        public static final byte CREATE_NEW_GUILD = 0x4;
        public static final byte INVITE_GUILD = 0x5;
        public static final byte JOIN_GUILD = 0x6;
        public static final byte WITHDRAW_GUILD = 0x7;
        public static final byte KICK_GUILD = 0x8;
        public static final byte REMOVE_GUILD = 0x9;
        public static final byte INC_MAX_MEMBER_NUM = 0xA;
        public static final byte CHANGE_LEVEL = 0xB;
        public static final byte CHANGE_JOB = 0xC;
        public static final byte SET_GRADE_NAME = 0xD;
        public static final byte SET_MEMBER_GRADE = 0xE;
        public static final byte SET_MARK = 0xF;
        public static final byte SET_NOTICE = 0x10;
        public static final byte INPUT_MARK = 0x11;
        public static final byte CHECK_QUEST_WAITING = 0x12;
        public static final byte CHECK_QUEST_WAITING_2 = 0x13;
        public static final byte INSERT_QUEST_WAITING = 0x14;
        public static final byte CANCEL_QUEST_WAITING = 0x15;
        public static final byte REMOVE_QUEST_COMPLETE_GUILD = 0x16;
        public static final byte INC_POINT = 0x17;
        public static final byte INC_COMMITMENT = 0x18;
        public static final byte SET_QUEST_TIME = 0x19;
        public static final byte SHOW_GUILD_RANKING = 0x1A;
        public static final byte SET_SKILL = 0x1B;
    }

    public static final class GuildRes {
        public static final byte LOAD_GUILD_DONE = 0x1C;
        public static final byte CHECK_GUILD_NAME_AVAILABLE = 0x1D;
        public static final byte CHECK_GUILD_NAME_ALREADY_USED = 0x1E;
        public static final byte CHECK_GUILD_NAME_UNKNOWN = 0x1F;
        public static final byte CREATE_GUILD_AGREE_REPLY = 0x20;
        public static final byte CREATE_GUILD_AGREE_UNKNOWN = 0x21;
        public static final byte CREATE_NEW_GUILD_DONE = 0x22;
        public static final byte CREATE_NEW_GUILD_ALREAY_JOINED = 0x23;
        public static final byte CREATE_NEW_GUILD_GUILD_NAME_ALREAY_EXIST = 0x24;
        public static final byte CREATE_NEW_GUILD_BEGINNER = 0x25;
        public static final byte CREATE_NEW_GUILD_DISAGREE = 0x26;
        public static final byte CREATE_NEW_GUILD_NOT_FULL_PARTY = 0x27;
        public static final byte CREATE_NEW_GUILD_UNKNOWN = 0x28;
        public static final byte JOIN_GUILD_DONE = 0x29;
        public static final byte JOIN_GUILD_ALREADY_JOINED = 0x2A;
        public static final byte JOIN_GUILD_ALREADY_FULL = 0x2B;
        public static final byte JOIN_GUILD_UNKNOWN_USER = 0x2C;
        public static final byte JOIN_GUILD_UNKNOWN = 0x2D;
        public static final byte WITHDRAW_GUILD_DONE = 0x2E;
        public static final byte WITHDRAW_GUILD_NOT_JOINED = 0x2F;
        public static final byte WITHDRAW_GUILD_UNKNOWN = 0x30;
        public static final byte KICK_GUILD_DONE = 0x31;
        public static final byte KICK_GUILD_NOT_JOINED = 0x32;
        public static final byte KICK_GUILD_UNKNOWN = 0x33;
        public static final byte REMOVE_GUILD_DONE = 0x34;
        public static final byte REMOVE_GUILD_NOT_EXIST = 0x35;
        public static final byte REMOVE_GUILD_UNKNOWN = 0x36;
        public static final byte INVITE_GUILD_BLOCKED_USER = 0x37;
        public static final byte INVITE_GUILD_ALREADY_INVITED = 0x38;
        public static final byte INVITE_GUILD_REJECTED = 0x39;
        public static final byte ADMIN_CANNOT_CREATE = 0x3A;
        public static final byte ADMIN_CANNOT_INVITE = 0x3B;
        public static final byte INC_MAX_MEMBER_NUM_DONE = 0x3C;
        public static final byte INC_MAX_MEMBER_NUM_UNKNOWN = 0x3D;
        public static final byte CHANGE_LEVEL_OR_JOB = 0x3E;
        public static final byte NOTIFY_LOGIN_OR_LOGOUT = 0x3F;
        public static final byte SET_GRADE_NAME_DONE = 0x40;
        public static final byte SET_GRADE_NAME_UNKNOWN = 0x41;
        public static final byte SET_MEMBER_GRADE_DONE = 0x42;
        public static final byte SET_MEMBER_GRADE_UNKNOWN = 0x43;
        public static final byte SET_MEMBER_COMMITMENT_DONE = 0x44;
        public static final byte SET_MARK_DONE = 0x45;
        public static final byte SET_MARK_UNKNOWN = 0x46;
        public static final byte SET_NOTICE_DONE = 0x47;
        public static final byte INSERT_QUEST = 0x48;
        public static final byte NOTICE_QUEST_WAITING_ORDER = 0x49;
        public static final byte SET_GUILD_CAN_ENTER_QUEST = 0x4A;
        public static final byte INC_POINT_DONE = 0x4B;
        public static final byte SHOW_GUILD_RANKING = 0x4C;
        public static final byte GUILD_QUEST_NOT_ENOUGH_USER = 0x4D;
        public static final byte GUILD_QUEST_REGISTER_DISCONNECTED = 0x4E;
        public static final byte GUILD_QUEST_NOTICE_ORDER = 0x4F;
        public static final byte AUTHKEY_UPDATE = 0x50;
        public static final byte SET_SKILL_DONE = 0x51;
        public static final byte SERVER_MSG = 0x52;
    }
}

File: src/main/java/net/maple/packets/GuildPackets.kt
package net.maple.packets

import client.Character
import net.database.GuildAPI
import net.maple.SendOpcode
import net.maple.packets.AlliancePackets.notifyLoginOrLogout
import util.packet.Packet
import util.packet.PacketWriter
import world.guild.Guild
import world.guild.GuildMark
import java.util.*

object GuildPackets {

    fun changeGuildName(chr: Character, name: String) {
        val pw = PacketWriter(8)

        pw.writeHeader(SendOpcode.USER_GUILD_NAME_CHANGED)
        pw.writeInt(chr.id)
        pw.writeMapleString(name)

        chr.field.broadcast(pw.createPacket(), chr)
    }

    fun inputMark(chr: Character) {
        val pw = PacketWriter(3)

        pw.writeHeader(SendOpcode.GUILD_RESULT)
        pw.writeByte(GuildReq.INPUT_MARK)

        chr.write(pw.createPacket())
    }

    fun Guild.getLoadGuildPacket(): Packet {
        val pw = PacketWriter(32)

        pw.writeHeader(SendOpcode.GUILD_RESULT)
        pw.writeByte(GuildRes.LOAD_GUILD_DONE)
        pw.writeBool(true)
        this.encode(pw)

        return pw.createPacket()
    }

    fun expel(chr: Character) {
        val pw = PacketWriter(4)

        pw.writeHeader(SendOpcode.GUILD_RESULT)
        pw.writeByte(GuildRes.LOAD_GUILD_DONE)
        pw.writeBool(false)

        chr.write(pw.createPacket())
    }

    fun Guild.getJoinGuildPacket(chr: Character): Packet {
        val pw = PacketWriter(32)

        pw.writeHeader(SendOpcode.GUILD_RESULT)
        pw.writeByte(GuildRes.JOIN_GUILD_DONE)
        pw.writeInt(this.id)
        pw.writeInt(chr.id)
        this.getMemberSecure(chr.id).encode(pw)

        return pw.createPacket()
    }

    fun Guild.sendInvite(invited: Character, inviter: Character) {
        val pw = PacketWriter(24)

        pw.writeHeader(SendOpcode.GUILD_RESULT)
        pw.writeByte(GuildReq.INVITE_GUILD)
        pw.writeInt(this.id)
        pw.writeMapleString(inviter.name)
        pw.writeInt(inviter.level)
        pw.writeInt(inviter.job)

        invited.write(pw.createPacket())
    }

    fun message(receiver: Character, message: Byte) {
        val pw = PacketWriter(3)

        pw.writeHeader(SendOpcode.GUILD_RESULT)
        pw.writeByte(message)

        receiver.write(pw.createPacket())
    }

    fun Guild.notifyLoginLogout(chr: Character, online: Boolean) {
        val pw = PacketWriter(12)

        pw.writeHeader(SendOpcode.GUILD_RESULT)
        pw.write(GuildRes.NOTIFY_LOGIN_OR_LOGOUT.toInt())
        pw.writeInt(id)
        pw.writeInt(chr.id)
        pw.writeBool(online)

        broadcast(pw.createPacket(), chr)
        alliance?.notifyLoginOrLogout(chr, online)
    }

    fun Guild.leave(cid: Int, name: String, message: Byte) {
        val pw = PacketWriter(16)

        pw.writeHeader(SendOpcode.GUILD_RESULT)
        pw.write(message.toInt())
        pw.writeInt(this.id)
        pw.writeInt(cid)
        pw.writeMapleString(name)

        this.broadcast(pw.createPacket())
    }

    fun Guild.setNotice(notice: String) {
        val pw = PacketWriter(16)

        pw.writeHeader(SendOpcode.GUILD_RESULT)
        pw.write(GuildRes.SET_NOTICE_DONE.toInt())
        pw.writeInt(this.id)
        pw.writeMapleString(notice)

        this.broadcast(pw.createPacket())
    }

    fun Guild.setMemberGrade(cid: Int, grade: Byte) {
        val pw = PacketWriter(12)

        pw.writeHeader(SendOpcode.GUILD_RESULT)
        pw.writeByte(GuildRes.SET_MEMBER_GRADE_DONE)
        pw.writeInt(this.id)
        pw.writeInt(cid)
        pw.writeByte(grade)

        this.broadcast(pw.createPacket())
    }

    fun Guild.setGradeNames() {
        val pw = PacketWriter(16)

        pw.writeHeader(SendOpcode.GUILD_RESULT)
        pw.writeByte(GuildRes.SET_GRADE_NAME_DONE)
        pw.writeInt(this.id)
        Arrays.stream(this.ranks).forEach { pw.writeMapleString(it ?: "") }

        this.broadcast(pw.createPacket())
    }

    fun removeGuild(guild: Guild) {
        val pw = PacketWriter(7)

        pw.writeHeader(SendOpcode.GUILD_RESULT)
        pw.writeByte(GuildRes.REMOVE_GUILD_DONE)
        pw.writeInt(guild.id)

        guild.broadcast(pw.createPacket())
    }

    fun Guild.increaseMemberSize(newSize: Byte) {
        val pw = PacketWriter(8)

        pw.writeHeader(SendOpcode.GUILD_RESULT)
        pw.writeByte(GuildRes.INC_MAX_MEMBER_NUM_DONE)
        pw.writeInt(id)
        pw.writeByte(newSize)

        broadcast(pw.createPacket())
        GuildAPI.updateMemberSize(this)
    }

    fun Guild.setGuildMarkPacket(): Packet {
        val pw = PacketWriter(13)

        pw.writeHeader(SendOpcode.GUILD_RESULT)
        pw.writeByte(GuildRes.SET_MARK_DONE)
        pw.writeInt(id)
        mark?.encode(pw) ?: pw.write(ByteArray(6))

        return pw.createPacket()
    }

    fun changeGuildMarkRemote(chr: Character, mark: GuildMark?) {
        val pw = PacketWriter(12)

        pw.writeHeader(SendOpcode.USER_GUILD_MARK_CHANGED)
        pw.writeInt(chr.id)

        mark?.encode(pw) ?: pw.write(ByteArray(6))

        chr.field.broadcast(pw.createPacket(), chr)
    }

    object GuildReq {
        const val LOAD_GUILD: Byte = 0x00
        const val INPUT_GUILD_NAME: Byte = 0x01
        const val CHECK_GUILD_NAME: Byte = 0x02
        const val CREATE_GUILD_AGREE: Byte = 0x03
        const val CREATE_NEW_GUILD: Byte = 0x04
        const val INVITE_GUILD: Byte = 0x05
        const val JOIN_GUILD: Byte = 0x06
        const val WITHDRAW_GUILD: Byte = 0x07
        const val KICK_GUILD: Byte = 0x08
        const val REMOVE_GUILD: Byte = 0x09
        const val INC_MAX_MEMBER_NUM: Byte = 0x0A
        const val CHANGE_LEVEL: Byte = 0x0B
        const val CHANGE_JOB: Byte = 0x0C
        const val SET_GRADE_NAME: Byte = 0x0D
        const val SET_MEMBER_GRADE: Byte = 0x0E
        const val SET_MARK: Byte = 0x0F
        const val SET_NOTICE: Byte = 0x10
        const val INPUT_MARK: Byte = 0x11
        const val CHECK_QUEST_WAITING: Byte = 0x12
        const val CHECK_QUEST_WAITING_2: Byte = 0x13
        const val INSERT_QUEST_WAITING: Byte = 0x14
        const val CANCEL_QUEST_WAITING: Byte = 0x15
        const val REMOVE_QUEST_COMPLETE_GUILD: Byte = 0x16
        const val INC_POINT: Byte = 0x17
        const val INC_COMMITMENT: Byte = 0x18
        const val SET_QUEST_TIME: Byte = 0x19
        const val SHOW_GUILD_RANKING: Byte = 0x1A
        const val SET_SKILL: Byte = 0x1B
    }

    object GuildRes {
        const val LOAD_GUILD_DONE: Byte = 0x1C
        const val CHECK_GUILD_NAME_AVAILABLE: Byte = 0x1D
        const val CHECK_GUILD_NAME_ALREADY_USED: Byte = 0x1E
        const val CHECK_GUILD_NAME_UNKNOWN: Byte = 0x1F
        const val CREATE_GUILD_AGREE_REPLY: Byte = 0x20
        const val CREATE_GUILD_AGREE_UNKNOWN: Byte = 0x21
        const val CREATE_NEW_GUILD_DONE: Byte = 0x22
        const val CREATE_NEW_GUILD_ALREAY_JOINED: Byte = 0x23
        const val CREATE_NEW_GUILD_GUILD_NAME_ALREAY_EXIST: Byte = 0x24
        const val CREATE_NEW_GUILD_BEGINNER: Byte = 0x25
        const val CREATE_NEW_GUILD_DISAGREE: Byte = 0x26
        const val CREATE_NEW_GUILD_NOT_FULL_PARTY: Byte = 0x27
        const val CREATE_NEW_GUILD_UNKNOWN: Byte = 0x28
        const val JOIN_GUILD_DONE: Byte = 0x29
        const val JOIN_GUILD_ALREADY_JOINED: Byte = 0x2A
        const val JOIN_GUILD_ALREADY_FULL: Byte = 0x2B
        const val JOIN_GUILD_UNKNOWN_USER: Byte = 0x2C
        const val JOIN_GUILD_UNKNOWN: Byte = 0x2D
        const val WITHDRAW_GUILD_DONE: Byte = 0x2E
        const val WITHDRAW_GUILD_NOT_JOINED: Byte = 0x2F
        const val WITHDRAW_GUILD_UNKNOWN: Byte = 0x30
        const val KICK_GUILD_DONE: Byte = 0x31
        const val KICK_GUILD_NOT_JOINED: Byte = 0x32
        const val KICK_GUILD_UNKNOWN: Byte = 0x33
        const val REMOVE_GUILD_DONE: Byte = 0x34
        const val REMOVE_GUILD_NOT_EXIST: Byte = 0x35
        const val REMOVE_GUILD_UNKNOWN: Byte = 0x36
        const val INVITE_GUILD_BLOCKED_USER: Byte = 0x37
        const val INVITE_GUILD_ALREADY_INVITED: Byte = 0x38
        const val INVITE_GUILD_REJECTED: Byte = 0x39
        const val ADMIN_CANNOT_CREATE: Byte = 0x3A
        const val ADMIN_CANNOT_INVITE: Byte = 0x3B
        const val INC_MAX_MEMBER_NUM_DONE: Byte = 0x3C
        const val INC_MAX_MEMBER_NUM_UNKNOWN: Byte = 0x3D
        const val CHANGE_LEVEL_OR_JOB: Byte = 0x3E
        const val NOTIFY_LOGIN_OR_LOGOUT: Byte = 0x3F
        const val SET_GRADE_NAME_DONE: Byte = 0x40
        const val SET_GRADE_NAME_UNKNOWN: Byte = 0x41
        const val SET_MEMBER_GRADE_DONE: Byte = 0x42
        const val SET_MEMBER_GRADE_UNKNOWN: Byte = 0x43
        const val SET_MEMBER_COMMITMENT_DONE: Byte = 0x44
        const val SET_MARK_DONE: Byte = 0x45
        const val SET_MARK_UNKNOWN: Byte = 0x46
        const val SET_NOTICE_DONE: Byte = 0x47
        const val INSERT_QUEST: Byte = 0x48
        const val NOTICE_QUEST_WAITING_ORDER: Byte = 0x49
        const val SET_GUILD_CAN_ENTER_QUEST: Byte = 0x4A
        const val INC_POINT_DONE: Byte = 0x4B
        const val SHOW_GUILD_RANKING: Byte = 0x4C
        const val GUILD_QUEST_NOT_ENOUGH_USER: Byte = 0x4D
        const val GUILD_QUEST_REGISTER_DISCONNECTED: Byte = 0x4E
        const val GUILD_QUEST_NOTICE_ORDER: Byte = 0x4F
        const val AUTHKEY_UPDATE: Byte = 0x50
        const val SET_SKILL_DONE: Byte = 0x51
        const val SERVER_MSG: Byte = 0x52
    }
}